components/datetime/src/format/datetime.rs

```diff
@@ -91,62 +91,68 @@ impl<'l> fmt::Display for FormattedDateTime<'l> {
 }
 
 // Apply length to input number and write to result using fixed_decimal_format.
-fn format_number<W>(
+fn try_write_number<W>(
     result: &mut W,
-    fixed_decimal_format: &FixedDecimalFormatter,
+    fixed_decimal_format: Option<&FixedDecimalFormatter>,
     mut num: FixedDecimal,
     length: FieldLength,
-) -> fmt::Result
+) -> Result<Result<(), Error>, fmt::Error>
 where
     W: fmt::Write + ?Sized,
 {
-    match length {
-        FieldLength::One | FieldLength::NumericOverride(_) => {}
-        FieldLength::TwoDigit => {
-            num.pad_start(2);
-            num.set_max_position(2);
-        }
-        FieldLength::Abbreviated => {
-            num.pad_start(3);
-        }
-        FieldLength::Wide => {
-            num.pad_start(4);
-        }
-        FieldLength::Narrow => {
-            num.pad_start(5);
-        }
-        FieldLength::Six => {
-            num.pad_start(6);
-        }
-        FieldLength::Fixed(p) => {
-            num.pad_start(p as i16);
-            num.set_max_position(p as i16);
+    if let Some(fdf) = fixed_decimal_format {
+        match length {
+            FieldLength::One | FieldLength::NumericOverride(_) => {}
+            FieldLength::TwoDigit => {
+                num.pad_start(2);
+                num.set_max_position(2);
+            }
+            FieldLength::Abbreviated => {
+                num.pad_start(3);
+            }
+            FieldLength::Wide => {
+                num.pad_start(4);
+            }
+            FieldLength::Narrow => {
+                num.pad_start(5);
+            }
+            FieldLength::Six => {
+                num.pad_start(6);
+            }
+            FieldLength::Fixed(p) => {
+                num.pad_start(p as i16);
+                num.set_max_position(p as i16);
+            }
         }
-    }
 
-    let formatted = fixed_decimal_format.format(&num);
-    formatted.write_to(result)
+        fdf.format(&num).write_to(result)?;
+        Ok(Ok(()))
+    } else {
+        num.write_to(result)?;
```

Comment:
Suggestion: This should write an Error part. While you're in here, maybe change all the `W: fmt::Write + ?Sized` to `PartsWrite + ?Sized`.